Python/pythonrun.c

```diff
@@ -1479,6 +1479,55 @@ PyOS_CheckStack(void)
 
 #endif /* WIN32 && _MSC_VER */
 
+#if defined(__APPLE__)
+
+#define MAC_OS_10_9 13
+#define MAC_OS_10_10 14
+#define MAC_OS_10_11 15
+
+#include <sys/utsname.h>
+
+/*
+ * Return non-zero when we run out of memory on the stack; zero otherwise.
+ */
+
+int
+PyOS_CheckStack(void)
+{
+    const pthread_t p_thread = pthread_self();
+    const uintptr_t end = (uintptr_t)pthread_get_stackaddr_np(p_thread);
+    const uintptr_t frame = (uintptr_t)__builtin_frame_address(0);
+    size_t stack_space = 0;
+
+    if (stack_space == 0) {
+        if (pthread_main_np() == 1) {
+            // On macOS 10.09 - 10.11 pthread_get_stacksize_np 
+            // returns wrong stack size on main thread.
+            // ref: https://github.com/rust-lang/rust/issues/43347#issuecomment-316783599
+            // ref: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8020753
+            struct utsname sysinfo;
+            uname(&sysinfo);
+            const int major_version = atoi(sysinfo.release);
```

Comment:
uname() is a syscall: it isn't free. Can't you cache this value in static variable, since it's not going to change in the process. Whereas PyOS_CheckStack() is called very often.